rfd/0001-runjob-design.md

```diff
@@ -0,0 +1,360 @@
+- If the process is launched successfully, the library will return an `io.ReadCloser` which can be read for the combined `STDOUT` and `STDERR` stream.
+
+There might be some other related details to work through while implementing this functionality, but generally speaking, this will be the process for executing jobs.
+
+### Job Output Streaming
+
+The library will stream the combined output of `STDOUT` and `STDERR` of running processes in an `io.ReadCloser` back to the server. The library and server will be connected to the same pipe created before the `jobmanager runjob` is called to start the requested process.
+
+The server will use a pub/sub model to distribute Jobs' output to clients. Each Job's output will be stored in its own associated buffer, which can then be replayed to one or many Subscribers (clients). The Jobs' buffer will support concurrent reads and writes via a mutex.  Subscribers can be created at any time and will be fed all of the contents in the buffer created since the start of the Job.
+
+No durable storage of jobs or job messages is supported. If the RunJob server crashes or restarts, all output and state of running processes are lost.
+
+### Job Shutdown
+
+In Linux, a process group can be treated like one process for signaling.  Each process group has a process group ID (PGID).
+
+Because the `command` process from `runjob` is getting its own PGID, we can send a signal to the process group so that it and its child processes can be shut down with one signal.  However, those child processes may have also created process groups.
+
+When a `Stop` command is requested for a Job, here are the steps needed to shut down all related processes for a given Job:
+
+- The `jobmanager` will call a mutex lock on the Job in the JobManager
+- The `jobmanager` will send a `SIGQUIT` to the `runjob` associated with the Job.
+- The `runjob` will get the PPID for it's child process via `syscall.Getpgid`, and send the `SIGQUIT` to that Process Group so that all of the child processes in the group will get the signal
+- After 10 seconds, if the `runjob` child process hasn't exited, it will then send a `SIGKILL` to the process group and then exit its own process and close the `io.ReadCloser`
+- Because of setting `SysProcAttr.Pdeathsig` when all of these associated processes were created, they should all get cleaned-up by the kernel if something happens like one of the child processes creating its own process group
+- Once the `io.ReadCloser` closes, the `jobmanager` will call the library to clean-up the cgroup associated with the job
+- The `jobmanager` will then release the mutex and mark the Job as competed
+
+### Security
+
+Plaintext connections will not be accepted.  Every connection to the service must use mTLS 1.3.
+
+All requests will require trusted and signed certificates.
+
+#### Authentication
+
+The server will require the client to use a certificate signed by a trusted authority.
```

Comment:
could you include a script or makefile that generates an example CA and client/server cert/key?

Reply:
Apologies for the ham-fisted manual edits to the args.  They should be fixed now after updating code/repasting everything that is working.

I also added a Makefile for generating and validating the certs, as well as added a bit in the design doc about how to generate them.   Lots more detail about them in the doc as well.